Sources/NIO/NonBlockingFileIO.swift

```diff
@@ -214,6 +214,54 @@ public struct NonBlockingFileIO {
         }
     }
 
+    /// Write `byteCount` bytes from `buffer` to `fileHandle in `NonBlockingFileIO`'s private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// The returned `ByteBuffer` will not have less than `byteCount` bytes unless we hit end-of-file in which
+    /// case the `ByteBuffer` will contain the bytes available to read.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `FileHandle` to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    ///   - byteCount: The number of bytes to write from `buffer` to `fileHandle`.
+    ///   - eventLoop: The `EventLoop` to create the returned `EventLoopFuture` from.
+    /// - returns: An `EventLoopFuture` which is fulfilled if the write was successful or fails on error.
+    public func write(fileHandle: FileHandle,
+                      buffer: ByteBuffer,
+                      byteCount: Int,
+                      eventLoop: EventLoop) -> EventLoopFuture<()> {
+        guard byteCount > 0 else {
+            return eventLoop.newSucceededFuture(result: ())
+        }
+
+        var buf = buffer
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            var bytesWritten = 0
+            while bytesWritten < byteCount {
+                let n = try buf.readWithUnsafeMutableReadableBytes { ptr in
+                    let res = try fileHandle.withUnsafeFileDescriptor { descriptor in
+                        try Posix.write(descriptor: descriptor,
+                                        pointer: ptr.baseAddress!,
+                                        size: byteCount - bytesWritten)
```

Comment:
We need some caution here too, as the `ptr` may not have `bytesCount - bytesWritten` bytes in it. If we add the check above to use `min(buffer.readableBytes, byteCount)` then we'll be safe, but we should probably still put a `precondition` that `ptr.count == byteCount` in this body just to be sure.